text/0000-fabric-private-chaincode.md

```diff
@@ -0,0 +1,284 @@
+ 
+ 
+## Principal Components of FPC:
+ 
+FPC Chaincode: A chaincode created by a developer to run in a Chaincode Enclave. Unlike regular Fabric chaincode, an Enclave Chaincode  must currently be written in C++ using our FPC SDK. A future goal for the project is to support additional languages, e.g., by the use of WebAssembly.
+
+Chaincode Enclave: An enclave in which a single particular chaincode executes.  The Chaincode Enclave contains the chaincode (called an Enclave Chaincode) to be executed, with its linked Chaincode Library. The FPC runtime is responsible for creating the Chaincode Enclave when the FPC chaincode is installed on the endorsing peers.
+ 
+Chaincode Library (also known as the FPC Shim): This is a shim interface exposed to an FPC Chaincode.  The Chaincode Library contains two parts, one residing inside the enclave and another one outside the enclave. The shim interface exposed to the FPC chaincode is written primarily in C++, whereas the counterpart is written in Go and mainly responsible for communication between the Peer and the FPC chaincode. The Chaincode Library enables an Enclave Chaincode to talk to the Peer, in particular to invoke the chaincode and provide "normal" shim operations such as getState and putState. The FPC SDK provides the chaincode library to enable FPC chaincode developers to easily link to this library when compiling their FPC Chaincode and building a deployment artifact. It implements the interface from the Peer to the Enclave Chaincode, encrypts and decrypts state and arguments, and implements Attestation.
+ 
+Ledger Enclave: The Ledger Enclave is a crucial component to protect the execution of a FPC chaincode. In particular, it is responsible to ensure that a FPC chaincode processes committed world state data, i.e. only state data that comes from valid and committed transactions. The Ledger Enclave is also called the TLCC, Trusted Ledger Enclave or the Validation Enclave. The Ledger Enclave is a separate enclave which locally stores integrity metadata for validating information on the blockchain ledger. Like the Peer, it performs standard validation logic when a new block arrives, but it also creates and stores a cryptographic hash of each new key-value pair of the blockchain state. This makes it possible to verify that the data coming from the Peer is correct -- remember that in the FPC setting we do not necessarily trust a (single) peer. The Chaincode Enclave uses the Ledger Enclave as its tamperproof source of information on the blockchain world state. The Ledger Enclave and a Chaincode Enclave interact by through a secure channel that is initially established when the Chaincode Enclave is deployed. There is one Ledger Enclave per Channel on any given Peer. When a peer receives a new block from the ordering service, it forwards it to the Ledger Enclave. The Ledger Enclave is part of the FPC runtime.
```

Comment:
Conceptionally, the ledger enclave maintains a hash of each kv pair in memory. As you pointed out, memory is a limited resources here and eventually the ledger enclave needs to implement a clever way to tackle this problem securely and efficiently. 

The ledger enclave can maintain the integrity metadata (hashes) in a merkle tree structure and outsource them securely (using enclave data sealing) to untrusted disk on the peer. 

Enclave data sealing provides authenticated encryption that enables the enclave to detect data corruption. In case of a violation, the ledger enclave must signal this violation to the chaincode that must abort transactions.

Reply:
Merkle tree-based integrity management is nice.

Reply:
Using a regular Merkle tree is not enough because you also need to authenticate non-membership queries, and there is no way to prove a key doesn't exist in a Merkle tree efficiently without the keys being sorted in some manner.